src/Npgsql/NpgsqlBatchCommand.cs

```diff
@@ -273,6 +275,62 @@ internal void ApplyCommandComplete(CommandCompleteMessage msg)
 
     internal void ResetPreparation() => ConnectorPreparedOn = null;
 
+    internal void PopulateOutputParameters(NpgsqlDataReader reader, ILogger logger)
+    {
+        Debug.Assert(_parameters is not null);
+        var parameters = _parameters;
+        var placeholderType = parameters.PlaceholderType;
+
+        // In the case of named and mixed parameters we first try to populate based on a named match.
+        // If we can't find the name we fill the foremost output direction parameter and increment our position.
+        // In positional and mixed modes we only populate by ordinal if the parameter is positional.
+
+        // For backwards compat we allow populating named parameters by ordinal as long as they haven't been filled yet.
+        // This means a row like {"a" => 1, "some_field" => 2} will populate the following output db params {"a" => 1, "b" => 2}.
+        // Not taking into account filled named parameters would otherwise result in these output db params {"a" => 2, "b" => null}.
+        HashSet<NpgsqlParameter>? filled = null;
+        var paramIndex = 0;
+        for (var i = 0; i < reader.FieldCount; i++)
+        {
+            if (placeholderType is PlaceholderType.Named or PlaceholderType.Mixed &&
+                parameters.TryGetValue(reader.GetName(i), out var p) && p.IsOutputDirection)
+            {
+                if (placeholderType is PlaceholderType.Named)
+                    (filled ??= new()).Add(p);
+                SetValue(reader, logger, p, i);
+            }
+            else
+            {
+                var parameterList = parameters.InternalList;
+                while (paramIndex < parameterList.Count)
+                {
+                    var idx = paramIndex++;
+                    if (parameterList[idx] is { IsOutputDirection: true } parameter
+                        && ((placeholderType is PlaceholderType.Named && filled?.Contains(parameter) != true)
+                                || parameter.IsPositional))
+                    {
+                        SetValue(reader, logger, parameter, i, idx);
+                        break;
+                    }
+                }
+            }
```

Comment:
Just to clarify, let's say we have named parameters 'a' and 'b' and in the data row it will be 'c' and 'a' in that particular order. Previously, we would have added 'c' to queue to be mapped to 'b' (because 'a' from data row is mapped to 'a' in parameters). But now, we'll immediately go below, check that the first parameter ('a') isn't filled (it isn't) and then we set it. Is it an intended behavour?

Reply:
Will check in a bit, probably not?

Reply:
Yeah I'm not sure what my thinking was here last year. As soon as the first column isn't matching a parameter name we get into trouble already. I do think there is a more optimal implementation of the original thing though.